File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/messages/SignalServiceDataMessage.java
/*
 * Copyright (C) 2014-2016 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */

package org.whispersystems.signalservice.api.messages;

import org.whispersystems.libsignal.InvalidMessageException;
import org.whispersystems.libsignal.util.guava.Optional;
import org.whispersystems.signalservice.api.messages.shared.SharedContact;
import org.whispersystems.signalservice.api.push.SignalServiceAddress;
import org.whispersystems.signalservice.api.util.OptionalUtil;

import java.util.LinkedList;
import java.util.List;
import java.util.UUID;

/**
 * Represents a decrypted Signal Service data message.
 */
public class SignalServiceDataMessage {

  private final long                                    timestamp;
  private final Optional<List<SignalServiceAttachment>> attachments;
  private final Optional<String>                        body;
  private final Optional<SignalServiceGroupContext>     group;
  private final Optional<byte[]>                        profileKey;
  private final boolean                                 endSession;
  private final boolean                                 expirationUpdate;
  private final int                                     expiresInSeconds;
  private final boolean                                 profileKeyUpdate;
  private final Optional<Quote>                         quote;
  private final Optional<List<SharedContact>>           contacts;
  private final Optional<List<Preview>>                 previews;
  private final Optional<List<Mention>>                 mentions;
  private final Optional<Sticker>                       sticker;
  private final boolean                                 viewOnce;
  private final Optional<Reaction>                      reaction;
  private final Optional<RemoteDelete>                  remoteDelete;
  private final Optional<GroupCallUpdate>               groupCallUpdate;

  /**
   * Construct a SignalServiceDataMessage.
   *
   * @param timestamp The sent timestamp.
   * @param group The group information (or null if none).
   * @param groupV2 The group information (or null if none).
   * @param attachments The attachments (or null if none).
   * @param body The message contents.
   * @param endSession Flag indicating whether this message should close a session.
   * @param expiresInSeconds Number of seconds in which the message should disappear after being seen.
   */
  SignalServiceDataMessage(long timestamp,
                           SignalServiceGroup group, SignalServiceGroupV2 groupV2,
                           List<SignalServiceAttachment> attachments,
                           String body, boolean endSession, int expiresInSeconds,
                           boolean expirationUpdate, byte[] profileKey, boolean profileKeyUpdate,
                           Quote quote, List<SharedContact> sharedContacts, List<Preview> previews,
                           List<Mention> mentions, Sticker sticker, boolean viewOnce, Reaction reaction, RemoteDelete remoteDelete,
                           GroupCallUpdate groupCallUpdate)
  {
    try {
      this.group = SignalServiceGroupContext.createOptional(group, groupV2);
    } catch (InvalidMessageException e) {
      throw new AssertionError(e);
    }

    this.timestamp        = timestamp;
    this.body             = OptionalUtil.absentIfEmpty(body);
    this.endSession       = endSession;
    this.expiresInSeconds = expiresInSeconds;
    this.expirationUpdate = expirationUpdate;
    this.profileKey       = Optional.fromNullable(profileKey);
    this.profileKeyUpdate = profileKeyUpdate;
    this.quote            = Optional.fromNullable(quote);
    this.sticker          = Optional.fromNullable(sticker);
    this.viewOnce         = viewOnce;
    this.reaction         = Optional.fromNullable(reaction);
    this.remoteDelete     = Optional.fromNullable(remoteDelete);
    this.groupCallUpdate  = Optional.fromNullable(groupCallUpdate);

    if (attachments != null && !attachments.isEmpty()) {
      this.attachments = Optional.of(attachments);
    } else {
      this.attachments = Optional.absent();
    }

    if (sharedContacts != null && !sharedContacts.isEmpty()) {
      this.contacts = Optional.of(sharedContacts);
    } else {
      this.contacts = Optional.absent();
    }

    if (previews != null && !previews.isEmpty()) {
      this.previews = Optional.of(previews);
    } else {
      this.previews = Optional.absent();
    }

    if (mentions != null && !mentions.isEmpty()) {
      this.mentions = Optional.of(mentions);
    } else {
      this.mentions = Optional.absent();
    }
  }

  public static Builder newBuilder() {
    return new Builder();
  }

  /**
   * @return The message timestamp.
   */
  public long getTimestamp() {
    return timestamp;
  }

  /**
   * @return The message attachments (if any).
   */
  public Optional<List<SignalServiceAttachment>> getAttachments() {
    return attachments;
  }

  /**
   * @return The message body (if any).
   */
  public Optional<String> getBody() {
    return body;
  }

  /**
   * @return The message group context (if any).
   */
  public Optional<SignalServiceGroupContext> getGroupContext() {
    return group;
  }

  public boolean isEndSession() {
    return endSession;
  }

  public boolean isExpirationUpdate() {
    return expirationUpdate;
  }

  public boolean isProfileKeyUpdate() {
    return profileKeyUpdate;
  }

  public boolean isGroupV1Update() {
    return group.isPresent() &&
           group.get().getGroupV1().isPresent() &&
           group.get().getGroupV1().get().getType() != SignalServiceGroup.Type.DELIVER;
  }

  public boolean isGroupV2Message() {
    return group.isPresent() &&
           group.get().getGroupV2().isPresent();
  }

  public boolean isGroupV2Update() {
    return isGroupV2Message() &&
           group.get().getGroupV2().get().hasSignedGroupChange() &&
           !hasRenderableContent();
  }

  public boolean isEmptyGroupV2Message() {
    return isGroupV2Message() && !isGroupV2Update() && !hasRenderableContent();
  }

  /** Contains some user data that affects the conversation */
  public boolean hasRenderableContent() {
    return attachments.isPresent()   ||
           body.isPresent()          ||
           quote.isPresent()         ||
           contacts.isPresent()      ||
           previews.isPresent()      ||
           mentions.isPresent()      ||
           sticker.isPresent()       ||
           reaction.isPresent()      ||
           remoteDelete.isPresent();
  }

  public int getExpiresInSeconds() {
    return expiresInSeconds;
  }

  public Optional<byte[]> getProfileKey() {
    return profileKey;
  }

  public Optional<Quote> getQuote() {
    return quote;
  }

  public Optional<List<SharedContact>> getSharedContacts() {
    return contacts;
  }

  public Optional<List<Preview>> getPreviews() {
    return previews;
  }

  public Optional<List<Mention>> getMentions() {
    return mentions;
  }

  public Optional<Sticker> getSticker() {
    return sticker;
  }

  public boolean isViewOnce() {
    return viewOnce;
  }

  public Optional<Reaction> getReaction() {
    return reaction;
  }

  public Optional<RemoteDelete> getRemoteDelete() {
    return remoteDelete;
  }

  public Optional<GroupCallUpdate> getGroupCallUpdate() {
    return groupCallUpdate;
  }

  public static class Builder {

    private List<SignalServiceAttachment> attachments    = new LinkedList<>();
    private List<SharedContact>           sharedContacts = new LinkedList<>();
    private List<Preview>                 previews       = new LinkedList<>();
    private List<Mention>                 mentions       = new LinkedList<>();

    private long                 timestamp;
    private SignalServiceGroup   group;
    private SignalServiceGroupV2 groupV2;
    private String               body;
    private boolean              endSession;
    private int                  expiresInSeconds;
    private boolean              expirationUpdate;
    private byte[]               profileKey;
    private boolean              profileKeyUpdate;
    private Quote                quote;
    private Sticker              sticker;
    private boolean              viewOnce;
    private Reaction             reaction;
    private RemoteDelete         remoteDelete;
    private GroupCallUpdate      groupCallUpdate;

    private Builder() {}

    public Builder withTimestamp(long timestamp) {
      this.timestamp = timestamp;
      return this;
    }

    public Builder asGroupMessage(SignalServiceGroup group) {
      if (this.groupV2 != null) {
        throw new AssertionError("Can not contain both V1 and V2 group contexts.");
      }
      this.group = group;
      return this;
    }

    public Builder asGroupMessage(SignalServiceGroupV2 group) {
      if (this.group != null) {
        throw new AssertionError("Can not contain both V1 and V2 group contexts.");
      }
      this.groupV2 = group;
      return this;
    }

    public Builder withAttachment(SignalServiceAttachment attachment) {
      this.attachments.add(attachment);
      return this;
    }

    public Builder withAttachments(List<SignalServiceAttachment> attachments) {
      this.attachments.addAll(attachments);
      return this;
    }

    public Builder withBody(String body) {
      this.body = body;
      return this;
    }

    public Builder asEndSessionMessage() {
      return asEndSessionMessage(true);
    }

    public Builder asEndSessionMessage(boolean endSession) {
      this.endSession = endSession;
      return this;
    }

    public Builder asExpirationUpdate() {
      return asExpirationUpdate(true);
    }

    public Builder asExpirationUpdate(boolean expirationUpdate) {
      this.expirationUpdate = expirationUpdate;
      return this;
    }

    public Builder withExpiration(int expiresInSeconds) {
      this.expiresInSeconds = expiresInSeconds;
      return this;
    }

    public Builder withProfileKey(byte[] profileKey) {
      this.profileKey = profileKey;
      return this;
    }

    public Builder asProfileKeyUpdate(boolean profileKeyUpdate) {
      this.profileKeyUpdate = profileKeyUpdate;
      return this;
    }

    public Builder withQuote(Quote quote) {
      this.quote = quote;
      return this;
    }

    public Builder withSharedContact(SharedContact contact) {
      this.sharedContacts.add(contact);
      return this;
    }

    public Builder withSharedContacts(List<SharedContact> contacts) {
      this.sharedContacts.addAll(contacts);
      return this;
    }

    public Builder withPreviews(List<Preview> previews) {
      this.previews.addAll(previews);
      return this;
    }

    public Builder withMentions(List<Mention> mentions) {
      this.mentions.addAll(mentions);
      return this;
    }

    public Builder withSticker(Sticker sticker) {
      this.sticker = sticker;
      return this;
    }

    public Builder withViewOnce(boolean viewOnce) {
      this.viewOnce = viewOnce;
      return this;
    }

    public Builder withReaction(Reaction reaction) {
      this.reaction = reaction;
      return this;
    }

    public Builder withRemoteDelete(RemoteDelete remoteDelete) {
      this.remoteDelete = remoteDelete;
      return this;
    }

    public Builder withGroupCallUpdate(GroupCallUpdate groupCallUpdate) {
      this.groupCallUpdate = groupCallUpdate;
      return this;
    }

    public SignalServiceDataMessage build() {
      if (timestamp == 0) timestamp = System.currentTimeMillis();
      return new SignalServiceDataMessage(timestamp, group, groupV2, attachments, body, endSession,
                                          expiresInSeconds, expirationUpdate, profileKey,
                                          profileKeyUpdate, quote, sharedContacts, previews,
                                          mentions, sticker, viewOnce, reaction, remoteDelete,
                                          groupCallUpdate);
    }
  }

  public static class Quote {
    private final long                   id;
    private final SignalServiceAddress   author;
    private final String                 text;
    private final List<QuotedAttachment> attachments;
    private final List<Mention>          mentions;

    public Quote(long id, SignalServiceAddress author, String text, List<QuotedAttachment> attachments, List<Mention> mentions) {
      this.id          = id;
      this.author      = author;
      this.text        = text;
      this.attachments = attachments;
      this.mentions    = mentions;
    }

    public long getId() {
      return id;
    }

    public SignalServiceAddress getAuthor() {
      return author;
    }

    public String getText() {
      return text;
    }

    public List<QuotedAttachment> getAttachments() {
      return attachments;
    }

    public List<Mention> getMentions() {
      return mentions;
    }

    public static class QuotedAttachment {
      private final String                  contentType;
      private final String                  fileName;
      private final SignalServiceAttachment thumbnail;

      public QuotedAttachment(String contentType, String fileName, SignalServiceAttachment thumbnail) {
        this.contentType = contentType;
        this.fileName    = fileName;
        this.thumbnail   = thumbnail;
      }

      public String getContentType() {
        return contentType;
      }

      public String getFileName() {
        return fileName;
      }

      public SignalServiceAttachment getThumbnail() {
        return thumbnail;
      }
    }
  }

  public static class Preview {
    private final String                            url;
    private final String                            title;
    private final String                            description;
    private final long                              date;
    private final Optional<SignalServiceAttachment> image;

    public Preview(String url, String title, String description, long date, Optional<SignalServiceAttachment> image) {
      this.url         = url;
      this.title       = title;
      this.description = description;
      this.date        = date;
      this.image       = image;
    }

    public String getUrl() {
      return url;
    }

    public String getTitle() {
      return title;
    }

    public String getDescription() {
      return description;
    }

    public long getDate() {
      return date;
    }

    public Optional<SignalServiceAttachment> getImage() {
      return image;
    }
  }

  public static class Sticker {
    private final byte[]                  packId;
    private final byte[]                  packKey;
    private final int                     stickerId;
    private final String                  emoji;
    private final SignalServiceAttachment attachment;

    public Sticker(byte[] packId, byte[] packKey, int stickerId, String emoji, SignalServiceAttachment attachment) {
      this.packId     = packId;
      this.packKey    = packKey;
      this.stickerId  = stickerId;
      this.emoji      = emoji;
      this.attachment = attachment;
    }

    public byte[] getPackId() {
      return packId;
    }

    public byte[] getPackKey() {
      return packKey;
    }

    public int getStickerId() {
      return stickerId;
    }

    public String getEmoji() {
      return emoji;
    }

    public SignalServiceAttachment getAttachment() {
      return attachment;
    }
  }

  public static class Reaction {
    private final String               emoji;
    private final boolean              remove;
    private final SignalServiceAddress targetAuthor;
    private final long                 targetSentTimestamp;

    public Reaction(String emoji, boolean remove, SignalServiceAddress targetAuthor, long targetSentTimestamp) {
      this.emoji               = emoji;
      this.remove              = remove;
      this.targetAuthor        = targetAuthor;
      this.targetSentTimestamp = targetSentTimestamp;
    }

    public String getEmoji() {
      return emoji;
    }

    public boolean isRemove() {
      return remove;
    }

    public SignalServiceAddress getTargetAuthor() {
      return targetAuthor;
    }

    public long getTargetSentTimestamp() {
      return targetSentTimestamp;
    }
  }

  public static class RemoteDelete {
    private final long targetSentTimestamp;

    public RemoteDelete(long targetSentTimestamp) {
      this.targetSentTimestamp = targetSentTimestamp;
    }

    public long getTargetSentTimestamp() {
      return targetSentTimestamp;
    }
  }

  public static class Mention {
    private final UUID uuid;
    private final int  start;
    private final int  length;

    public Mention(UUID uuid, int start, int length) {
      this.uuid   = uuid;
      this.start  = start;
      this.length = length;
    }

    public UUID getUuid() {
      return uuid;
    }

    public int getStart() {
      return start;
    }

    public int getLength() {
      return length;
    }
  }

  public static class GroupCallUpdate {
    private final String eraId;

    public GroupCallUpdate(String eraId) {
      this.eraId = eraId;
    }

    public String getEraId() {
      return eraId;
    }
  }
}

File: libsignal-service/src/main/java/org/whispersystems/signalservice/api/messages/SignalServiceDataMessage.kt
/*
 * Copyright (C) 2014-2016 Open Whisper Systems
 *
 * Licensed according to the LICENSE file in this repository.
 */
package org.whispersystems.signalservice.api.messages

import org.signal.libsignal.zkgroup.groups.GroupSecretParams
import org.signal.libsignal.zkgroup.receipts.ReceiptCredentialPresentation
import org.whispersystems.signalservice.api.messages.shared.SharedContact
import org.whispersystems.signalservice.api.push.ServiceId
import org.whispersystems.signalservice.api.util.OptionalUtil.asOptional
import org.whispersystems.signalservice.api.util.OptionalUtil.emptyIfStringEmpty
import org.whispersystems.signalservice.internal.push.BodyRange
import java.util.LinkedList
import java.util.Optional
import org.whispersystems.signalservice.internal.push.DataMessage.Payment as PaymentProto
import org.whispersystems.signalservice.internal.push.DataMessage.Quote as QuoteProto

/**
 * Represents a decrypted Signal Service data message.
 *
 * @param timestamp The sent timestamp.
 * @param groupContext The group information (or null if none).
 * @param attachments The attachments (or null if none).
 * @param body The message contents.
 * @param isEndSession Flag indicating whether this message should close a session.
 * @param expiresInSeconds Number of seconds in which the message should disappear after being seen.
 */
class SignalServiceDataMessage private constructor(
  val timestamp: Long,
  val groupContext: Optional<SignalServiceGroupV2>,
  val attachments: Optional<List<SignalServiceAttachment>>,
  val body: Optional<String>,
  val isEndSession: Boolean,
  val expiresInSeconds: Int,
  val isExpirationUpdate: Boolean,
  val profileKey: Optional<ByteArray>,
  val isProfileKeyUpdate: Boolean,
  val quote: Optional<Quote>,
  val sharedContacts: Optional<List<SharedContact>>,
  val previews: Optional<List<SignalServicePreview>>,
  val mentions: Optional<List<Mention>>,
  val sticker: Optional<Sticker>,
  val isViewOnce: Boolean,
  val reaction: Optional<Reaction>,
  val remoteDelete: Optional<RemoteDelete>,
  val groupCallUpdate: Optional<GroupCallUpdate>,
  val payment: Optional<Payment>,
  val storyContext: Optional<StoryContext>,
  val giftBadge: Optional<GiftBadge>,
  val bodyRanges: Optional<List<BodyRange>>
) {
  val isActivatePaymentsRequest: Boolean = payment.map { it.isActivationRequest }.orElse(false)
  val isPaymentsActivated: Boolean = payment.map { it.isActivation }.orElse(false)

  val groupId: Optional<ByteArray> = groupContext.map { GroupSecretParams.deriveFromMasterKey(it.masterKey).publicParams.groupIdentifier.serialize() }
  val isGroupV2Message: Boolean = groupContext.isPresent

  /** Contains some user data that affects the conversation  */
  private val hasRenderableContent: Boolean =
    this.attachments.isPresent ||
      this.body.isPresent ||
      this.quote.isPresent ||
      this.sharedContacts.isPresent ||
      this.previews.isPresent ||
      this.mentions.isPresent ||
      this.sticker.isPresent ||
      this.reaction.isPresent ||
      this.remoteDelete.isPresent

  val isGroupV2Update: Boolean = groupContext.isPresent && groupContext.get().hasSignedGroupChange() && !hasRenderableContent
  val isEmptyGroupV2Message: Boolean = isGroupV2Message && !isGroupV2Update && !hasRenderableContent

  class Builder {
    private var timestamp: Long = 0
    private var groupV2: SignalServiceGroupV2? = null
    private val attachments: MutableList<SignalServiceAttachment> = LinkedList<SignalServiceAttachment>()
    private var body: String? = null
    private var endSession: Boolean = false
    private var expiresInSeconds: Int = 0
    private var expirationUpdate: Boolean = false
    private var profileKey: ByteArray? = null
    private var profileKeyUpdate: Boolean = false
    private var quote: Quote? = null
    private val sharedContacts: MutableList<SharedContact> = LinkedList<SharedContact>()
    private val previews: MutableList<SignalServicePreview> = LinkedList<SignalServicePreview>()
    private val mentions: MutableList<Mention> = LinkedList<Mention>()
    private var sticker: Sticker? = null
    private var viewOnce: Boolean = false
    private var reaction: Reaction? = null
    private var remoteDelete: RemoteDelete? = null
    private var groupCallUpdate: GroupCallUpdate? = null
    private var payment: Payment? = null
    private var storyContext: StoryContext? = null
    private var giftBadge: GiftBadge? = null
    private var bodyRanges: MutableList<BodyRange> = LinkedList<BodyRange>()

    fun withTimestamp(timestamp: Long): Builder {
      this.timestamp = timestamp
      return this
    }

    fun asGroupMessage(group: SignalServiceGroupV2?): Builder {
      groupV2 = group
      return this
    }

    fun withAttachments(attachments: List<SignalServiceAttachment>?): Builder {
      attachments?.let { this.attachments.addAll(attachments) }
      return this
    }

    fun withBody(body: String?): Builder {
      this.body = body
      return this
    }

    @JvmOverloads
    fun asEndSessionMessage(endSession: Boolean = true): Builder {
      this.endSession = endSession
      return this
    }

    @JvmOverloads
    fun asExpirationUpdate(expirationUpdate: Boolean = true): Builder {
      this.expirationUpdate = expirationUpdate
      return this
    }

    fun withExpiration(expiresInSeconds: Int): Builder {
      this.expiresInSeconds = expiresInSeconds
      return this
    }

    fun withProfileKey(profileKey: ByteArray?): Builder {
      this.profileKey = profileKey
      return this
    }

    fun asProfileKeyUpdate(profileKeyUpdate: Boolean): Builder {
      this.profileKeyUpdate = profileKeyUpdate
      return this
    }

    fun withQuote(quote: Quote?): Builder {
      this.quote = quote
      return this
    }

    fun withSharedContact(contact: SharedContact?): Builder {
      contact?.let { sharedContacts.add(contact) }
      return this
    }

    fun withSharedContacts(contacts: List<SharedContact>?): Builder {
      contacts?.let { sharedContacts.addAll(contacts) }
      return this
    }

    fun withPreviews(previews: List<SignalServicePreview>?): Builder {
      previews?.let { this.previews.addAll(previews) }
      return this
    }

    fun withMentions(mentions: List<Mention>?): Builder {
      mentions?.let { this.mentions.addAll(mentions) }
      return this
    }

    fun withSticker(sticker: Sticker?): Builder {
      this.sticker = sticker
      return this
    }

    fun withViewOnce(viewOnce: Boolean): Builder {
      this.viewOnce = viewOnce
      return this
    }

    fun withReaction(reaction: Reaction?): Builder {
      this.reaction = reaction
      return this
    }

    fun withRemoteDelete(remoteDelete: RemoteDelete?): Builder {
      this.remoteDelete = remoteDelete
      return this
    }

    fun withGroupCallUpdate(groupCallUpdate: GroupCallUpdate?): Builder {
      this.groupCallUpdate = groupCallUpdate
      return this
    }

    fun withPayment(payment: Payment?): Builder {
      this.payment = payment
      return this
    }

    fun withStoryContext(storyContext: StoryContext?): Builder {
      this.storyContext = storyContext
      return this
    }

    fun withGiftBadge(giftBadge: GiftBadge?): Builder {
      this.giftBadge = giftBadge
      return this
    }

    fun withBodyRanges(bodyRanges: List<BodyRange>?): Builder {
      bodyRanges?.let { this.bodyRanges.addAll(bodyRanges) }
      return this
    }

    fun build(): SignalServiceDataMessage {
      if (timestamp == 0L) {
        timestamp = System.currentTimeMillis()
      }

      return SignalServiceDataMessage(
        timestamp = timestamp,
        groupContext = groupV2.asOptional(),
        attachments = attachments.asOptional(),
        body = body.emptyIfStringEmpty(),
        isEndSession = endSession,
        expiresInSeconds = expiresInSeconds,
        isExpirationUpdate = expirationUpdate,
        profileKey = profileKey.asOptional(),
        isProfileKeyUpdate = profileKeyUpdate,
        quote = quote.asOptional(),
        sharedContacts = sharedContacts.asOptional(),
        previews = previews.asOptional(),
        mentions = mentions.asOptional(),
        sticker = sticker.asOptional(),
        isViewOnce = viewOnce,
        reaction = reaction.asOptional(),
        remoteDelete = remoteDelete.asOptional(),
        groupCallUpdate = groupCallUpdate.asOptional(),
        payment = payment.asOptional(),
        storyContext = storyContext.asOptional(),
        giftBadge = giftBadge.asOptional(),
        bodyRanges = bodyRanges.asOptional()
      )
    }
  }

  data class Quote(
    val id: Long,
    val author: ServiceId?,
    val text: String,
    val attachments: List<QuotedAttachment>?,
    val mentions: List<Mention>?,
    val type: Type,
    val bodyRanges: List<BodyRange>?
  ) {
    enum class Type(val protoType: QuoteProto.Type) {
      NORMAL(QuoteProto.Type.NORMAL),
      GIFT_BADGE(QuoteProto.Type.GIFT_BADGE);

      companion object {
        @JvmStatic
        fun fromProto(protoType: QuoteProto.Type): Type {
          return values().firstOrNull { it.protoType == protoType } ?: NORMAL
        }
      }
    }

    data class QuotedAttachment(val contentType: String, val fileName: String?, val thumbnail: SignalServiceAttachment?)
  }
  class Sticker(val packId: ByteArray?, val packKey: ByteArray?, val stickerId: Int, val emoji: String?, val attachment: SignalServiceAttachment?)
  data class Reaction(val emoji: String, val isRemove: Boolean, val targetAuthor: ServiceId, val targetSentTimestamp: Long)
  data class RemoteDelete(val targetSentTimestamp: Long)
  data class Mention(val serviceId: ServiceId, val start: Int, val length: Int)
  data class GroupCallUpdate(val eraId: String?)
  class PaymentNotification(val receipt: ByteArray, val note: String)
  data class PaymentActivation(val type: PaymentProto.Activation.Type)
  class Payment(paymentNotification: PaymentNotification?, paymentActivation: PaymentActivation?) {
    val paymentNotification: Optional<PaymentNotification> = Optional.ofNullable(paymentNotification)
    val paymentActivation: Optional<PaymentActivation> = Optional.ofNullable(paymentActivation)
    val isActivationRequest: Boolean = paymentActivation != null && paymentActivation.type == PaymentProto.Activation.Type.REQUEST
    val isActivation: Boolean = paymentActivation != null && paymentActivation.type == PaymentProto.Activation.Type.ACTIVATED
  }
  data class StoryContext(val authorServiceId: ServiceId, val sentTimestamp: Long)
  data class GiftBadge(val receiptCredentialPresentation: ReceiptCredentialPresentation)

  companion object {
    @JvmStatic
    fun newBuilder(): Builder {
      return Builder()
    }
  }
}
